### DISCUSSION TITLE
How to enable DOOR pausing?

### DISCUSSION THREAD

**Discussion by stevempotter** [phil-barrett/grblHAL-teensy-4.x#39] [Q&A]
I have uncommented line 86 in config.h
#define ENABLE_SAFETY_DOOR_INPUT_PIN 
and also uncommented the delays for spindle and coolant re-starting.
I recompiled and flashed the Teensy.
I have wired a NO switch (closes when door is open or E-stop button is hit) 
to the DOOR input on the Teensy BoB.
Nothing happens when I hit the switch during a run. It just keeps on going.
What other steps do I need to implement the DOOR feature?

**Comment by phil-barrett** [comment]
I did the same.  Default input pin - ie NC switch.  Seems to work. Here's a screenshot.  You can see multiple check door messages. The ss was taken with the input switch open. (edit, replaced SS with a slightly larger so you can see all the console info)
![image](https://user-images.githubusercontent.com/16570958/146239029-154b351e-5d2a-423e-96db-33e8fed184c8.png)

**Reply by phil-barrett** [reply]
Are you sure your switch is working (look at voltage at the door sig terminal for closed and open)?  Can you see if the output of the opto is getting pulled down when the switch closes?

**Comment by stevempotter** [comment]
1st problem: There is no D red square indicator in my ioSender screen. How to enable that?
2nd (solved): As you suspected, the switch must have been dirty (or else one of the wires was not inserted in the screw terminal correctly): the voltage only drifted slowly down to about 0.6 volts on pressing the button. I have re-installed the wires carefully and cleaned the switch contacts, and now it works well: goes right down to zero volts instantly when pressed, and up to 4.6V when released.

But still the job keeps going when the button is pressed. 

Do I need to set some things in ioSender such as to invert the NO switch's signal and to enable the D "LED"?

**Comment by stevempotter** [comment]
BTW, I have 2.0.36p14 version of ioSender.

**Comment by phil-barrett** [comment]
There isn't anything special you need to do beyond configuring it. If you aren't seeing the red box in ioSender then grblHAL isn't seeing the pin. Do you have any other plugins enabled?  Have you modified the map file?

You could also trace the signal input.  The schematic is in the user manual.  I would check to see if the teensy is getting the signal.

**Comment by stevempotter** [comment]
I verified that the hardware is OK: when no Teensy is plugged in, Pin29 goes from infinite to low resistance (to ground) when the button is pushed. But when the Teensy is plugged in to the BoB, that pin is always low. I think it should be being pulled high, and go low when the button is pressed right?
So I infer that I have failed to get my changes in config.h onto the Teensy. I am now trying to figure out how to compile everything in VS Code instead of Arduino. Building and reflashing this thing is not very easy! In my first attempts, I now seem to have broken the ethernet link. There are no more lights blinking on the ethernet jacks on either end.

**Reply by phil-barrett** [reply]
sigh, yes it can be tricky to get it all working.  Try juggling 10 different configurations! The arduino IDE is a mess.  

Glad your HW is ok.  Yes, the output of the opto (and input to the Teensy itself) on the BoB should be pulled high when the input switch is open and then low when it is closed.  For NC switches (grblHAL default), triggered (open) should read high.  As long as you are seeing a transition upon switch closure, it's not a HW problem.

Send me your my_machine.h and config.h so I can try building/testing it here.

**Reply by stevempotter** [reply]
OK, I sent them to your email...  Thanks

**Reply by phil-barrett** [reply]
Got them.  Rebuilt using both and got a .hex that works including door functionality.  Attached a zip of the .hex for you to try.
One thing I noticed - you have eeprom enabled.  Do you have one soldered on the board?  If you don't, comment out the EEPROM_ENABLE line. 

Anyway, try the .hex file and see if door works for you. If it doesn't we'll have to look at what else you have connected.

[grblHAL_Teensy41-smp.zip](https://github.com/phil-barrett/grblHAL-teensy-4.x/files/7731070/grblHAL_Teensy41-smp.zip)

**Comment by stevempotter** [comment]
Thanks, Phil. I think I successfully uploaded your .hex by dragging it onto the teensy.exe GUI and pushing the reset button. The verbose readout indicated blocks being written. But on plugging the Teensy back into the BoB, I still have a dead ethernet connection. No blinking ethernet LEDs, no connection to ioSender. So I tried connecting via a USB cable instead. ioSender's GUI opens with most tabs greyed out and no info in the console (see screenshot). Any ideas on how to debug this?
![ioSender interface - most tabs disabled](https://user-images.githubusercontent.com/5586932/146591654-5853f7d1-0ccd-4308-a26a-a63bfb725f63.jpg)
Thanks for the note about eeprom. I have re-commented it out. I had assumed it was referring to eeprom on the Teensy.

**Reply by phil-barrett** [reply]
That screenshot shows that ioSender has not actually connected to the Teensy. It should probably just say that...  Pretty sure you have the EEProm enabled - that tends to cause grblHAL to crash when the chip isn't actually there.

Here's a build with EEProm removed.

[grblHAL_TEENSY41_smp.zip](https://github.com/phil-barrett/grblHAL-teensy-4.x/files/7737228/grblHAL_TEENSY41_smp.zip)

**Comment by stevempotter** [comment]
Thanks, Phil. that .hex has brought the ethernet connection back to life, yay! And I see that Teensy Pin29 is now being pulled high, and goes low when the E-Stop button (Door input) is pressed. However, the ioSender interface still does not show the D "LED" indicator or the letter D next to the other endstops' indicators. There must be some setting in ioSender I must change to activate it?

**Comment by stevempotter** [comment]
More problems, I can't even get to the grbl settings. It seems my endstop signals are set wrong (because they came from your build?) and can't change them because it just keeps asking me if I want to end ioSender. I can't "clear" them all to move forward. This dialog should give the option of going to grbl settings to change them.
See attached screenshots.
![Cant get to grbl settings](https://user-images.githubusercontent.com/5586932/146647770-63c50521-d39d-4a66-a664-fc9b7b9aa40e.JPG)

![ioSender error msg](https://user-images.githubusercontent.com/5586932/146647779-bb72e549-8862-4a0b-a133-a482a06ae299.jpg)
![estops active false alarms](https://user-images.githubusercontent.com/5586932/146648031-df3f4dbe-1e0a-4a58-8252-c83f30150c13.jpg)

**Comment by phil-barrett** [comment]
OK, Alarm 10 is because grblHAL defaults to NC switches - E-Stop included. Try closing the halt switch input to at least get up and running.  [Read about that here.](https://github.com/terjeio/grblHAL/wiki/First-Run-Grbl-Settings) Also, I have a blog entry that [talks about the issues in a different context](https://www.grbl.org/single-post/naked-teensy-party).

By the way, the settings didn't come from my build, they are the default. Sometimes reloading firmware for a different build wipes out the previous settings.  I have a note to myself to work through the fine points of that, they aren't totally obvious - perhaps another blog.  The build I sent should have door enabled.  I'm running with it on my test setup and ioSender shows door (D).  Try getting past the Alarm 10/E-Stop issue.  I think this is another newbie friendliness issue.

**Reply by stevempotter** [reply]
That blog entry is very helpful, thanks, Phil. It is all working now. I was able to flash my own VSCode version of my firmware. I had to make sure when I opened the firmware folder in VSCode, that I was at the right point in the file hierarchy (at GRBLHAL_TEENSY4) to include the file platform.ini.  There were problems with the compiler complaining that I was supposed to be using TimeLib.h instead of time.h. I removed the #warning dummy version of time.h and then it compiled OK. What is all that about? 
Once I got my own version of firmware flashed, I no longer had these Alarm 10 warnings. BTW, I am using all hall-effect switches instead of microswitches for endstops, so they are powered and immune to noise.  I had to invert my DOOR pin (it is NO) in grbl settings (Control signals, Invert control pins), and configure the Safety Door to enable parking (pulling up the bit 10 mm). It all seems to work great now. 

I never would have guessed that enabling EEPROM would cause these problems that disabled my ethernet connection. The comments in my_machine.h really ought to be edited to include a warning that this option is only for those who have soldered on an EXTRA EEPROM chip. I had assumed I needed EEPROM to save all the settings in non-volatile memory.

**Reply by stevempotter** [reply]
> What is all that about?

From the PJRC forum: 
__
This is an Arduino problem not a SdFat problem.

For some reason the Arduino people accepted the TimeLib library with inside a Time.h file, that now only calls TimeLib.h (and gives the warning you reported). There is a toolchain level include file 'time.h' which on linux os has a distinct filename but not on Windows systems.
Here you see that 'time.h' is called from toolchain, but Arduino is finding 'Time.h' and includes the wrong file.

Best is to remove the file 'Time.h' inside the TimeLib library.
Only side-effect is that some 'wrong-written' Arduino libraries may need to be modified to include 'TimeLib.h' and not 'Time.h'. But this can be handled from case to case.
___

**Comment by phil-barrett** [comment]
Such is the way of open source!

Glad you got it all sorted out.   Good point about EEPROM.
